src/neuralnet/neuron_layer/gru.cc

```diff
@@ -0,0 +1,265 @@
+  delete new_memory;
+}
+
+void GRULayer::Setup(const LayerProto& conf,
+    const vector<Layer*>& srclayers) {
+  Layer::Setup(conf, srclayers);
+  CHECK_LE(srclayers.size(), 2);
+  const auto& src = srclayers[0]->data(this);
+  
+  batchsize_ = src.shape()[0]; // size of batch
+  vdim_ = src.count() / (batchsize_); // dimension of input
+
+  hdim_ = layer_conf_.gru_conf().dim_hidden(); // dimension of hidden state
+
+  data_.Reshape(vector<int>{batchsize_, hdim_});
+  grad_.ReshapeLike(data_);
+
+  // Initialize the parameters
+  weight_z_hx_ = Param::Create(conf.param(0));
+  weight_z_hh_ = Param::Create(conf.param(1));
+  bias_z_ = Param::Create(conf.param(2));
+
+  weight_r_hx_ = Param::Create(conf.param(3));
+  weight_r_hh_ = Param::Create(conf.param(4));
+  bias_r_ = Param::Create(conf.param(5)); 
+
+  weight_c_hx_ = Param::Create(conf.param(6));
+  weight_c_hh_ = Param::Create(conf.param(7));
+  bias_c_ = Param::Create(conf.param(8));
+  
+  
+  weight_z_hx_->Setup(vector<int>{hdim_, vdim_});
+  weight_z_hh_->Setup(vector<int>{hdim_, hdim_});
+  bias_z_->Setup(vector<int>{hdim_});
+
+  weight_r_hx_->Setup(vector<int>{hdim_, vdim_});
+  weight_r_hh_->Setup(vector<int>{hdim_, hdim_});
+  bias_r_->Setup(vector<int>{hdim_});
+
+  weight_c_hx_->Setup(vector<int>{hdim_, vdim_});
+  weight_c_hh_->Setup(vector<int>{hdim_, hdim_});
+  bias_c_->Setup(vector<int>{hdim_});
+
+  update_gate = new Blob<float>(batchsize_, hdim_);
+  reset_gate = new Blob<float>(batchsize_, hdim_);
+  new_memory = new Blob<float>(batchsize_, hdim_);
+
+}
+
+void GRULayer::ComputeFeature(int flag,
+    const vector<Layer*>& srclayers) {
+	CHECK_LE(srclayers.size(), 2);
+
+	// Do transpose
+	Blob<float> *w_z_hx_t = Transpose (weight_z_hx_->data());
+	Blob<float> *w_z_hh_t = Transpose (weight_z_hh_->data());
+	Blob<float> *w_r_hx_t = Transpose (weight_r_hx_->data());
+	Blob<float> *w_r_hh_t = Transpose (weight_r_hh_->data());
+	Blob<float> *w_c_hx_t = Transpose (weight_c_hx_->data());
+	Blob<float> *w_c_hh_t = Transpose (weight_c_hh_->data());
+
+	// Prepare the data input and the context
+	const auto& src = srclayers[0]->data(this);
+	const Blob<float> *context;
+	if (srclayers.size() == 1) { // only have data input
+		context = new Blob<float>(batchsize_, hdim_);
+	} else { // have data input & context
+		context = &srclayers[1]->data(this);
+	}
+
+	// Compute the update gate
+	GEMM(cpu, 1.0f, 0.0f, src,*w_z_hx_t,update_gate);
+	MVAddRow(cpu,1.0f,1.0f,bias_z_->data(),update_gate);
+	Blob<float> zprev (batchsize_,hdim_);
+	GEMM(cpu, 1.0f, 0.0f, *context,*w_z_hh_t, &zprev);
+	Add<float>(cpu, *update_gate, zprev, update_gate);
+	Map<op::Sigmoid<float>,float>(cpu, *update_gate, update_gate);
+
+	// Compute the reset gate
+	GEMM(cpu, 1.0f, 0.0f, src,*w_r_hx_t,reset_gate);
+	MVAddRow(cpu,1.0f,1.0f,bias_r_->data(),reset_gate);
+	Blob<float> rprev (batchsize_, hdim_);
+	GEMM(cpu, 1.0f, 0.0f, *context, *w_r_hh_t, &rprev);
+	Add<float>(cpu, *reset_gate, rprev, reset_gate);
```

Comment:
I checked this again and find it may not be merged, as the elements in "update_gate" should be initially set to 0 - that is why I used GEMM(cpu,1.0f,0.0f,....)...